File: packages/shared-schemas/src/shared_schemas/spec_review.py
from datetime import datetime
from enum import StrEnum
from uuid import UUID

from pydantic import BaseModel

from .common import BaseSchema


class ReviewIssueSeverity(StrEnum):
    critical = "critical"
    warning = "warning"
    info = "info"


class ReviewIssueCategory(StrEnum):
    contradiction = "contradiction"
    gap = "gap"
    ambiguity = "ambiguity"
    inconsistency = "inconsistency"


class SpecReviewIssue(BaseSchema):
    severity: ReviewIssueSeverity
    category: ReviewIssueCategory
    title: str
    description: str
    location: str | None = None  # Section or line reference in the spec


class SpecReviewSuggestion(BaseSchema):
    severity: ReviewIssueSeverity
    title: str
    description: str
    before: str | None = None  # Original text to replace
    after: str | None = None  # Proposed replacement text
    section: str | None = None  # Section to add/modify
    status: str = "pending"  # pending / applied / dismissed
    related_issue_index: int | None = None  # Index into issues list


class SpecReviewQuestion(BaseSchema):
    question: str
    context: str | None = None  # Why this question matters
    priority: str = "medium"  # high / medium / low


class SpecReviewSummary(BaseSchema):
    quality_score: int  # 0-100
    quality_badge: str  # excellent / good / needs_improvement / poor
    total_issues: int
    total_suggestions: int
    total_questions: int
    issues_by_category: dict[str, int]
    issues_by_severity: dict[str, int]


class SpecReviewRead(BaseSchema):
    id: UUID
    specification_id: UUID
    conversation_id: UUID
    job_id: str | None = None
    status: str
    scope: str
    summary: SpecReviewSummary | None = None
    issues: list[SpecReviewIssue] = []
    suggestions: list[SpecReviewSuggestion] = []
    questions: list[SpecReviewQuestion] = []
    created_at: datetime
    updated_at: datetime


class SpecReviewTriggerRequest(BaseModel):
    scope: str = "full"  # full / quick


class SpecReviewTriggerResponse(BaseSchema):
    job_id: str
    specification_id: UUID
    conversation_id: UUID
    message: str


class SuggestionApplyRequest(BaseModel):
    confirm: bool = False  # Set True to apply, False for preview only


class SuggestionApplyResponse(BaseSchema):
    preview_diff: str | None = None
    applied: bool = False
    updated_content: str | None = None  # Only when applied=True

File: apps/web/src/types/spec-review.ts
// TypeScript mirrors of shared_schemas.spec_review

export type ReviewIssueSeverity = "critical" | "warning" | "info";
export type ReviewIssueCategory =
  | "contradiction"
  | "gap"
  | "ambiguity"
  | "inconsistency";

export interface SpecReviewIssue {
  severity: ReviewIssueSeverity;
  category: ReviewIssueCategory;
  title: string;
  description: string;
  location: string | null;
}

export interface SpecReviewSuggestion {
  severity: ReviewIssueSeverity;
  title: string;
  description: string;
  before: string | null;
  after: string | null;
  section: string | null;
  status: "pending" | "applied" | "dismissed";
  related_issue_index: number | null;
}

export interface SpecReviewQuestion {
  question: string;
  context: string | null;
  priority: "high" | "medium" | "low";
}

export interface SpecReviewSummary {
  quality_score: number;
  quality_badge: string;
  total_issues: number;
  total_suggestions: number;
  total_questions: number;
  issues_by_category: Record<string, number>;
  issues_by_severity: Record<string, number>;
}

export interface SpecReviewRead {
  id: string;
  specification_id: string;
  conversation_id: string;
  job_id: string | null;
  status: string;
  scope: string;
  summary: SpecReviewSummary | null;
  issues: SpecReviewIssue[];
  suggestions: SpecReviewSuggestion[];
  questions: SpecReviewQuestion[];
  created_at: string;
  updated_at: string;
}

export interface SpecReviewTriggerResponse {
  job_id: string;
  specification_id: string;
  conversation_id: string;
  message: string;
}

export interface SuggestionApplyResponse {
  preview_diff: string | null;
  applied: boolean;
  updated_content: string | null;
}
